.github/workflows/failure-notification.yml

```diff
@@ -0,0 +1,88 @@
+# **********************************************************
+# Copyright (c) 2025 Google, Inc.  All rights reserved.
+# **********************************************************
+
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice,
+#   this list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of Google, Inc. nor the names of its contributors may be
+#   used to endorse or promote products derived from this software without
+#   specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+# DAMAGE.
+
+# Github Actions workflow to send out failure notification emails.
+# This is a reusable workflow that is called by various CI workflows.
+
+name: send-failure-notification
+on:
+  workflow_call:
+    inputs:
+      test_suite_status:
+        required: true
+        type: string
+      # We ask for the results only string separately so we can do a
+      # substring match to check for failure.
+      test_suite_results_only:
+        required: true
+        type: string
+    secrets:
+      DYNAMORIO_NOTIFICATION_EMAIL_USERNAME:
+        required: true
+      DYNAMORIO_NOTIFICATION_EMAIL_PASSWORD:
+        required: true
+
+jobs:
+  send-failure-notification-job:
+    runs-on: ubuntu-latest
+    # Reusable workflows are invoked as a separate job, which requires
+    # a new runner. But since this job is run only on master merge
+    # failures, which are expected to be rare, it is fine.
+    #
+    # We keep this condition in this reusable workflow to avoid its
+    # duplication across all callers. Callers may want to specify an
```

Comment:
Should "may" be stronger?  See other comment: I don't understand when it's skipped.  Don't you have deps on all the jobs via `needs`?

Reply:
> Should "may" be stronger? 

Did s/may want to/should/

> I don't understand when it's skipped.

`needs` specifies the jobs that must finish for this job to run. If any needed job fails, this job doesn't run.

> Don't you have deps on all the jobs via needs?

I don't understand: are you saying that the fact that a job has dependencies should somehow guarantee that it itself would be run?

Yes the failure notification  job depends on the test jobs. And Github Actions makes it so that a failing needed job, skips the dependent job.

Reply:
> If any needed job fails, this job doesn't run.

That's what I was missing: how about putting that in parens after "prematurely skipped".

Reply:
Done